fms/models/gpt_oss.py

```diff
@@ -0,0 +1,637 @@
+            device=self.embedding.weight.device,
+        )
+
+        layers = []
+        for i in range(self.config.nlayers):
+            config = self.config
+            if config.layer_types[i] == "full_attention":  # type: ignore[index]
+                config = self.config.updated(sliding_window=0)
+            block: nn.Module = GptOssBlock(config, self.rot_emb)
+            block = self.distributed_strategy.distribute_layer(block, i)
+            layers.append(block)
+        self.layers = nn.ModuleList(layers)
+
+        dec_norm = LayerNormParameterized(
+            config.emb_dim,
+            elementwise_scale=True,
+            elementwise_shift=False,
+            use_mean=False,
+            eps=config.norm_eps,
+            use_high_precision_pow=True,
+        )
+
+        self.dec_norm = self.distributed_strategy.distribute_module(
+            dec_norm, final_layers=True
+        )
+
+        if self.config.p_dropout:
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+    def reset_parameters(self):
+        nn.init.trunc_normal_(
+            self.embedding.weight, mean=0.0, std=self.config.emb_dim**-0.5
+        )
+
+        # Call reset_parameters for relevant sub-layers
+        for m in self.modules():
+            if isinstance(m, MultiHeadAttention) or isinstance(m, MOEFeedForward):
+                m.reset_parameters()
+
+    def post_init(self):
+        # This function is called in `get_model` after the model is
+        # fully initalized on the correct device
+        for device in set(
+            [param.device for param in self.parameters()]
+            + [buffer.device for buffer in self.buffers()]
+        ):
+            self.rot_emb.device = device
+
+    def forward(
+        self,
+        x_in,
+        position_ids=None,
+        past_key_value_states=None,
+        use_cache=False,
+        **attn_kwargs: Unpack[AttentionKwargs],
+    ):
+        # Embed the given vocabulary indices using the given attention mask, with pre-/post-norm and dropout as specified
+        # x_in: batch_size x seq_len x emb_dim if input is already embedded, otherwise batch_size x seq_len
+        # mask: batch_size x seq_len x seq_len
+        # bias: nheads x seq_len x seq_len
+        attn_kwargs.update({"attn_name": "sdpa_with_sinks"})
+
+        if past_key_value_states is None or len(past_key_value_states) == 0:
+            past_key_value_states = [None for _ in range(len(self.layers))]
+
+        if x_in.dim() == 2:  # input is not already embedded
+            x_in = self.embedding(x_in)
+
+        # this is the output cache for all the decoder layers
+        present_key_value_states = []
+
+        for i, layer in enumerate(self.layers):
+            if layer.config.sliding_window > 0:  # type: ignore
+                attn_kwargs.update({"sliding_window": layer.config.sliding_window})  # type: ignore
+            else:
+                attn_kwargs.update({"sliding_window": 0})
```

Comment:
instead of using update here, use the more pythonic `attn_kwargs["sliding_window"] = config.sliding_window`